clients/roscpp/src/libros/io.cpp

```diff
@@ -87,6 +94,69 @@ bool last_socket_error_is_would_block() {
 #endif
 }
 
+int create_socket_watcher() {
+  int epfd = -1;
+#if !defined(WIN32)
+  epfd = epoll_create1(0);
+  if (epfd < 0)
+  {
+    ROS_ERROR("Unable to create epoll watcher: %s", strerror(errno));
+  }
+#endif
+  return epfd;
+}
+
+void close_socket_watcher(int fd) {
+  if (fd >= 0)
+    close(fd);
+}
+
+void add_socket_to_watcher(int epfd, int fd) {
+#if defined(WIN32)
+  UNUSED(epfd);
+  UNUSED(fd);
+#else
+  struct epoll_event ev;
+  ev.events = 0;
+  ev.data.fd = fd;
+
+  if (epoll_ctl(epfd, EPOLL_CTL_ADD, fd, &ev))
+  {
+    ROS_ERROR("Unable to add FD to epoll: %s", strerror(errno));
+  }
+#endif
```

Comment:
I haven't fully digested all of what's going on here, but this if/else stuff going on everywhere feels like it's going to be a big pain to maintain. Should we be contemplating supplying multiple distinct implementations (epoll, poll, windows), and using a single branch somewhere to select which one is used?